ambry-frontend/src/main/java/com.github.ambry.frontend/AmbrySecurityService.java

```diff
@@ -341,4 +354,22 @@ private void setAccountAndContainerHeaders(RestRequest restRequest, RestResponse
     }
     restResponseChannel.setHeader(RestUtils.Headers.PRIVATE, !container.isCacheable());
   }
+
+  /**
+   * Validate the cdn signed path in the URI if required for specific {@link Container}.
+   * @param restRequest the {@link RestRequest} that may contain signed path.
+   * @param expectCdnSignedPath the expected cnd signed path specified in {@link com.github.ambry.config.FrontendConfig}.
+   * @throws RestServiceException
+   */
+  private void validateCdnSignedPathIfRequired(RestRequest restRequest, String expectCdnSignedPath)
+      throws RestServiceException {
+    Container targetContainer = getContainerFromArgs(restRequest.getArgs());
+    if (targetContainer.isCdnSignedPathValidationRequired()) {
+      String signedPathInUri = getCdnSignedPath(restRequest);
+      if (!signedPathInUri.equals(expectCdnSignedPath)) {
```

Comment:
After reconsideration, I think we should be strict on signed path validation in open source code. That is, if URI contains pure signed id or blob id (no path at the beginning), we still reject it because the empty path doesn't match expected signed path. Rejecting all mismatched URIs is more aligned with initial intention of introducing signed path validation for certain container. We can make extra changes in closed source code to differentiate external and internal URIs if needed.